diff -urN GotoBLAS~/common_power.h GotoBLAS/common_power.h
--- GotoBLAS~/common_power.h	2007-01-16 22:25:41.000000000 -0600
+++ GotoBLAS/common_power.h	2007-02-11 13:54:58.000000000 -0600
@@ -591,7 +591,7 @@
 #if defined(PPC440)
 #define BUFFER_SIZE     (  2 << 20)
 #elif defined(PPC440FP2)
-#define BUFFER_SIZE     ( 64 << 20)
+#define BUFFER_SIZE     (  4 << 20)
 #else
 #define BUFFER_SIZE     ( 64 << 20)
 #endif
diff -urN GotoBLAS~/getarch.c GotoBLAS/getarch.c
--- GotoBLAS~/getarch.c	2007-01-16 22:25:41.000000000 -0600
+++ GotoBLAS/getarch.c	2007-02-15 13:09:49.731705751 -0600
@@ -39,7 +39,7 @@
 #undef	FORCE_PPC970
 #undef	FORCE_PPC970MP
 #undef	FORCE_PPC440
-#undef	FORCE_PPC440FP2
+#define FORCE_PPC440FP2
 #undef	FORCE_CELL
 #undef	FORCE_CLEARSPEED
 #undef	FORCE_GENERIC
Files GotoBLAS~/getarch_cmd and GotoBLAS/getarch_cmd differ
Files GotoBLAS~/level1/others/libmisc.a and GotoBLAS/level1/others/libmisc.a differ
diff -urN GotoBLAS~/level1/others/memory.c GotoBLAS/level1/others/memory.c
--- GotoBLAS~/level1/others/memory.c	2007-01-16 22:25:41.000000000 -0600
+++ GotoBLAS/level1/others/memory.c	2007-02-11 13:41:33.000000000 -0600
@@ -55,7 +55,7 @@
 #include <sys/sysinfo.h>
 #include <sched.h>
 #include <errno.h>
-#include <linux/unistd.h>
+#include <unistd.h>
 #if 0
 #ifdef USE_MEMORY_AFFINITY
 #include <linux/mempolicy.h>
diff -urN GotoBLAS~/Makefile.rule GotoBLAS/Makefile.rule
--- GotoBLAS~/Makefile.rule	2007-01-16 22:25:41.000000000 -0600
+++ GotoBLAS/Makefile.rule	2007-02-11 13:41:33.000000000 -0600
@@ -7,7 +7,7 @@
 
 # Which do you prefer to use for C compiler? Default is gcc.
 # I recommend you to use GCC because inline assembler is required.
-# C_COMPILER = GNU
+C_COMPILER = GNU
 # C_COMPILER = INTEL
 
 # Which do you prefer to use for Fortran compiler? Default is GNU g77.
@@ -17,7 +17,7 @@
 # F_COMPILER = INTEL
 # F_COMPILER = PGI
 # F_COMPILER = PATHSCALE
-# F_COMPILER = IBM
+F_COMPILER = IBM
 # F_COMPILER = COMPAQ
 # F_COMPILER = SUN
 # F_COMPILER = F2C
@@ -34,7 +34,7 @@
 # You may specify Maximum number of threads. It should be minimum.
 # For OpenMP users; you have to specify MAX_THREADS even if you don't
 # use threaded BLAS(MAX_THREADS >= OMP_NUM_THREADS * GOTO_NUM_THREADS)
-# MAX_THREADS = 4
+MAX_THREADS = 1
 
 # Default setting is using busy wait synchronization instead of
 # pthread barrier to improve performance. If you see performance
@@ -66,6 +66,9 @@
 # thread overhead. This is not implemented yet.
 # CCOMMON_OPT	+= -DALL_THREADED
 
+CCOMMON_OPT += -Dlinux
+FCOMMON_OPT += -WF,-Dlinux
+
 # If you have special compiler to run script to determine architecture.
 GETARCH_CC	 =
 GETARCH_FLAGS	 =
@@ -145,7 +148,7 @@
 endif
 
 ifeq ($(C_COMPILER), GNU)
-COMPILER     = gcc
+COMPILER     = /bgl/BlueLight/ppcfloor/blrts-gnu/bin/powerpc-bgl-blrts-gnu-gcc
 CCOMMON_OPT += -Wall # -m128bit-long-double
 ifneq ($(OSNAME), CYGWIN_NT)
 CCOMMON_OPT += -fPIC 
@@ -240,9 +243,9 @@
 endif
 
 ifeq ($(F_COMPILER), IBM)
-COMPILER_F77 = xlf
+COMPILER_F77 = blrts_xlf
 BU	     = 
-# FCOMMON_OPT	+= -qarch=440
+FCOMMON_OPT	+= -qarch=440
 ifdef BINARY64
 FCOMMON_OPT += -q64
 ifdef INTERFACE64
